internal/manifest/manifest.go

```diff
@@ -180,3 +245,100 @@ func iteratePrefix[T prefixable](manifest *Manifest, prefix *T, onMatch func(*T)
 	}
 	return nil
 }
+
+func manifestAddPackages(dbw *jsonwall.DBWriter, infos []*archive.PackageInfo) error {
+	for _, info := range infos {
+		err := dbw.Add(&Package{
+			Kind:    "package",
+			Name:    info.Name,
+			Version: info.Version,
+			Digest:  info.SHA256,
+			Arch:    info.Arch,
+		})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func manifestAddSlices(dbw *jsonwall.DBWriter, slices []*setup.Slice) error {
+	for _, slice := range slices {
+		err := dbw.Add(&Slice{
+			Kind: "slice",
+			Name: slice.String(),
+		})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func manifestAddReport(dbw *jsonwall.DBWriter, entries map[string]ReportEntry) error {
+	for _, entry := range entries {
+		sliceNames := []string{}
+		for slice := range entry.Slices {
+			err := dbw.Add(&Content{
+				Kind:  "content",
+				Slice: slice.String(),
+				Path:  entry.Path,
+			})
+			if err != nil {
+				return err
+			}
+			sliceNames = append(sliceNames, slice.String())
+		}
+		sort.Strings(sliceNames)
+		err := dbw.Add(&Path{
+			Kind:      "path",
+			Path:      entry.Path,
+			Mode:      fmt.Sprintf("0%o", unixPerm(entry.Mode)),
+			Slices:    sliceNames,
+			Hash:      entry.Hash,
+			FinalHash: entry.FinalHash,
+			Size:      uint64(entry.Size),
+			Link:      entry.Link,
+		})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func manifestAddManifestPaths(dbw *jsonwall.DBWriter, manifestMode os.FileMode, manifestSlices map[string][]*setup.Slice) error {
+	for path, slices := range manifestSlices {
+		sliceNames := []string{}
+		for _, slice := range slices {
+			err := dbw.Add(&Content{
+				Kind:  "content",
+				Slice: slice.String(),
+				Path:  path,
+			})
+			if err != nil {
+				return err
+			}
+			sliceNames = append(sliceNames, slice.String())
+		}
+		sort.Strings(sliceNames)
+		err := dbw.Add(&Path{
+			Kind:   "path",
+			Path:   path,
+			Mode:   fmt.Sprintf("0%o", unixPerm(manifestMode)),
```

Comment:
How do we know this mode?  The answer is we don't, not here. That's a relatively small issue, but it's also a hint of a larger issue that is quite clear: there's an unfortunate cross-dependency and disconnect right now between the manifest package and the slicer package about how to write stuff to disk.

Let's please try to fix this by:

1. Introduce fsutil.CreateWriter, which returns (io.WriterCloser, *Entry, error). Let's please do it right, in the sense that we'll have a writerProxy, similar to the readerProxy that we have today, to fill up the hash and size as we go, as that's trivial to do. We won't depend on it, though. See below.

2. Add the *Entry to the actual Report as soon as we create the *Entry.  At this point the entry will still be empty as far as the hash goes, and that's the way it should be as otherwise we'd need to add it at the end, which doesn't work since entries are sorted.

3. Remove manifestAddManifestPaths and the respective field from the options type.

With these steps, there's nothing special anymore about manifest files here: the slicer cuts it, the report holds it, the manifest writes it down, just like everything else.

Reply:
One extra detail here: CreateWriter can take the same Options parameter we have in the Create function.

Reply:
Done in https://github.com/canonical/chisel/pull/142/commits/dad0ed70f65931c840662304fb1f4f861b6383ce.